### PR TITLE
Bump gh actions mac runners (gets us native M1 builds)

### PR DESCRIPTION
**Pull request by ankith26**
The `macos-14` runner runs on the M1 chips, now we can run all our builds natively without any cross compilation workarounds. This means that now we can also finally test our arm64 mac releases on CI

On the intel mac builds this should give a little speedup, as the macos-13 runner apparently has 4 cores instead of the 3 cores on macos 11/12 runners

### CODE REVIEW

**Review by MyreMylar — APPROVED** [review]
Looks good, makes sense to me to be running on native M1 now.

**Comment by ankith26** [comment]
I have updated this PR to make it even simpler to approve, by leaving out a change that tried to modify parallel job usage (it was not giving a significant improvement anyways)

**Comment by cclauss** [comment]
Over the next 12 weeks `macos-latest` will become `macos-14`... https://github.blog/changelog/2024-04-01-macos-14-sonoma-is-generally-available-and-the-latest-macos-runner-image

**Comment by ankith26** [comment]
Had to do a couple of updates to get this working again, as the runner imaged changed significantly since I made this PR. This PR is ready for review now, though.

File: .github/workflows/build-macos.yml [hunk 1]
```diff
@@ -38,11 +38,13 @@ concurrency:
 jobs:
   deps:
     name: ${{ matrix.macarch }} deps
-    runs-on: macos-12
+    runs-on: ${{ matrix.runner }}
```
[the remark this answers is not in the document]
Comment (ankith26) [reply .github/workflows/build-macos.yml]: Have renamed the variable to avoid the confusion

**Comment by cclauss** [comment]
This pull request has been open for 7+ months.  `.github/workflows/build-macos.yml` still only tests on `macos-12`

* #2775 was closed in favor of this pull request but it was moving in the right direction.

This pull request attempts to make a single large and complex home run.  Instead, it might be more effective to take an ___incremental approach of creating smaller PRs that are easier to reason about, review, and merge___.  At a minimum, we should already have one or more GitHub Actions jobs running on `macos-13` (Intel x64) and one or more jobs that run on `macos-14` (Apple SIlicon ARM64).  In less than a month macOS 15 will be announced so our upgrades should keep moving.

**Comment by ankith26** [comment]
I would also have liked to do that, but the thing is all these changes are inherently linked. If I split up the changes into parts, the parts may individually start to fail.

And this PR isn't that big to begin with. It makes testing a one-time thing and also easier

**Review by oddbookworm — APPROVED** [review]
LGTM!

**Pull request merged**